File: cache_interceptor/src/main/java/uk/co/glass_software/android/cache_interceptor/interceptors/error/ApiErrorFactory.java
package uk.co.glass_software.android.cache_interceptor.interceptors.error;

import android.support.annotation.NonNull;

import com.google.gson.JsonParseException;
import com.google.gson.stream.MalformedJsonException;

import java.io.IOException;

import retrofit2.HttpException;
import uk.co.glass_software.android.cache_interceptor.utils.Function;

import static uk.co.glass_software.android.cache_interceptor.interceptors.error.ApiError.NON_HTTP_STATUS;
import static uk.co.glass_software.android.cache_interceptor.interceptors.error.ErrorCode.NETWORK;
import static uk.co.glass_software.android.cache_interceptor.interceptors.error.ErrorCode.NOT_FOUND;
import static uk.co.glass_software.android.cache_interceptor.interceptors.error.ErrorCode.UNAUTHORISED;
import static uk.co.glass_software.android.cache_interceptor.interceptors.error.ErrorCode.UNEXPECTED_RESPONSE;
import static uk.co.glass_software.android.cache_interceptor.interceptors.error.ErrorCode.UNKNOWN;

public class ApiErrorFactory implements Function<Throwable, ApiError> {
    
    @Override
    public ApiError get(Throwable throwable) {
        if (throwable instanceof IOException || throwable instanceof JsonParseException) {
            return getIoError(throwable);
        }
        else if (throwable instanceof HttpException) {
            return getHttpError((HttpException) throwable);
        }
        else {
            ApiError apiError = ApiError.from(throwable);
            
            if (apiError == null) {
                return getError(
                        NON_HTTP_STATUS,
                        UNKNOWN,
                        throwable.getClass().getName()
                        + ": "
                        + throwable.getMessage(),
                        throwable
                );
            }
            
            return apiError;
        }
    }
    
    @NonNull
    private ApiError getHttpError(HttpException throwable) {
        ErrorCode errorCode = parseErrorCode(throwable);
        
        return getError(throwable.code(),
                        errorCode,
                        throwable.message(),
                        throwable
        );
    }
    
    @NonNull
    private ApiError getIoError(Throwable throwable) {
        if (throwable instanceof MalformedJsonException
            || throwable instanceof JsonParseException) {
            return getError(NON_HTTP_STATUS,
                            UNEXPECTED_RESPONSE,
                            throwable.getMessage(),
                            throwable
            );
        }
        else {
            return getError(NON_HTTP_STATUS,
                            NETWORK,
                            throwable.getMessage(),
                            throwable
            );
        }
    }
    
    @NonNull
    private ApiError getError(int httpStatus,
                              ErrorCode errorCode,
                              String rawDescription,
                              Throwable cause) {
        return new ApiError(
                cause,
                httpStatus,
                errorCode,
                rawDescription
        );
    }
    
    private ErrorCode parseErrorCode(HttpException httpException) {
        switch (httpException.code()) {
            case 401:
                return UNAUTHORISED;
            
            case 404:
                return NOT_FOUND;
            
            default:
                return UNKNOWN;
        }
    }
}

File: cache_interceptor/src/main/java/uk/co/glass_software/android/cache_interceptor/interceptors/error/ApiErrorFactory.kt
package uk.co.glass_software.android.cache_interceptor.interceptors.error

import com.google.gson.JsonParseException
import com.google.gson.stream.MalformedJsonException
import retrofit2.HttpException
import uk.co.glass_software.android.cache_interceptor.interceptors.error.ApiError.Companion.NON_HTTP_STATUS
import uk.co.glass_software.android.cache_interceptor.interceptors.error.ErrorCode.*
import java.io.IOException
import java.util.concurrent.TimeoutException

class ApiErrorFactory : (Throwable) -> ApiError {

    override fun invoke(throwable: Throwable) =
            if (throwable is IOException
                    || throwable is JsonParseException
                    || throwable is TimeoutException)
                getIoError(throwable)
            else if (throwable is HttpException)
                getHttpError(throwable)
            else
                ApiError.from(throwable) ?: getError(
                        NON_HTTP_STATUS,
                        UNKNOWN,
                        throwable.javaClass.name
                                + ": "
                                + throwable.message,
                        throwable
                )

    private fun getHttpError(throwable: HttpException) = getError(
            throwable.code(),
            parseErrorCode(throwable),
            throwable.message(),
            throwable
    )

    private fun getIoError(throwable: Throwable) =
            (throwable is MalformedJsonException || throwable is JsonParseException).let {
                getError(
                        NON_HTTP_STATUS,
                        if (it) UNEXPECTED_RESPONSE else NETWORK,
                        throwable.message,
                        throwable
                )
            }

    private fun getError(httpStatus: Int,
                         errorCode: ErrorCode,
                         rawDescription: String?,
                         cause: Throwable) = ApiError(
            cause,
            httpStatus,
            errorCode,
            rawDescription
    )

    private fun parseErrorCode(httpException: HttpException) =
            when (httpException.code()) {
                401 -> UNAUTHORISED
                404 -> NOT_FOUND
                else -> UNKNOWN
            }
}
